installinator/src/block_size_writer.rs

```diff
@@ -0,0 +1,223 @@
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use pin_project_lite::pin_project;
+use std::io;
+use std::pin::Pin;
+use std::task::ready;
+use std::task::Context;
+use std::task::Poll;
+use tokio::io::AsyncWrite;
+
+pin_project! {
+    /// `BlockSizeWriter` is analogous to a tokio's `BufWriter`, except it
+    /// guarantees that writes made to the underlying writer are always
+    /// _exactly_ the requested block size, with two exceptions: explicitly
+    /// calling (1) `flush()` or (2) `shutdown()` will write any
+    /// buffered-but-not-yet-written data to the underlying buffer regardless of
+    /// its length.
+    pub(crate) struct BlockSizeWriter<W> {
+        #[pin]
+        inner: W,
+        buf: Vec<u8>,
+        block_size: usize,
+    }
+}
+
+impl<W: AsyncWrite> BlockSizeWriter<W> {
+    pub(crate) fn with_block_size(block_size: usize, inner: W) -> Self {
+        Self { inner, buf: Vec::with_capacity(block_size), block_size }
+    }
+
+    #[cfg(test)]
+    fn into_inner(self) -> W {
+        self.inner
+    }
+
+    fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut W> {
+        self.project().inner
+    }
+
+    fn flush_buf(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<io::Result<()>> {
+        let mut me = self.project();
+        let mut written = 0;
+        let mut ret = Ok(());
+        while written < me.buf.len() {
+            match ready!(me.inner.as_mut().poll_write(cx, &me.buf[written..])) {
+                Ok(0) => {
+                    ret = Err(io::Error::new(
+                        io::ErrorKind::WriteZero,
+                        "failed to write the buffered data",
+                    ));
+                    break;
+                }
+                Ok(n) => written += n,
+                Err(e) => {
+                    ret = Err(e);
+                    break;
+                }
+            }
+        }
+        if written > 0 {
+            me.buf.drain(..written);
+        }
+        Poll::Ready(ret)
+    }
+}
+
+impl<W: AsyncWrite> AsyncWrite for BlockSizeWriter<W> {
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, std::io::Error>> {
+        // We should never buffer up more than `block_size` data.
+        assert!(self.buf.len() <= self.block_size);
+
+        // If we already have exactly `block_size` bytes, begin by flushing.
+        if self.buf.len() == self.block_size {
+            ready!(self.as_mut().flush_buf(cx))?;
+        }
+
+        let me = self.project();
+        if me.buf.is_empty() {
+            // If our buffer is empty, either directly write `block_size` bytes
+            // from `buf` to `inner` (if there's enough data in `buf`) or just
+            // copy it into our buffer.
+            if buf.len() >= *me.block_size {
```

Comment:
I think it may be clearer to use `==` here, since we explicitly want `buf.len() <= self.block_size` and we assert that above.

Reply:
Instead of renaming I suggest I do a better job reviewing :) Let's keep it as is. The naming is idiomatic.